lib/kafka/consumer.rb

```diff
@@ -452,24 +452,15 @@ def make_final_offsets_commit!(attempts = 3)
     end
 
     def join_group
-      old_generation_id = @group.generation_id
-
       @group.join
 
-      if old_generation_id && @group.generation_id != old_generation_id + 1
-        # We've been out of the group for at least an entire generation, no
-        # sense in trying to hold on to offset data
-        clear_current_offsets
-        @offset_manager.clear_offsets
-      else
-        # After rejoining the group we may have been assigned a new set of
-        # partitions. Keeping the old offset commits around forever would risk
-        # having the consumer go back and reprocess messages if it's assigned
-        # a partition it used to be assigned to way back. For that reason, we
-        # only keep commits for the partitions that we're still assigned.
-        clear_current_offsets(excluding: @group.assigned_partitions)
-        @offset_manager.clear_offsets_excluding(@group.assigned_partitions)
-      end
+      # we observed duplicates after rebalance because
+      # consumer reads old offsets from its buffer
+      # that's why we clear buffer to make it fetch offsets from Kafka
+      # moreover,
+      # native Java client always gets offsets from Kafka
+      clear_current_offsets
+      @offset_manager.clear_offsets
```

Comment:
I would add in a log statement here indicating that offsets were cleared. 
This would aid in investigation production issues.

Reply:
@DylanSD we have logs about rebalance, leave and join group, also we see where we start to read. I think that should be enough. these buffers are internal ruby-kafka feature, native clients don't have it, I don't know if it makes sense to log operations with them.